sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -247,3 +218,37 @@ trait HashJoin {
     }
   }
 }
+
+object HashJoin {
+  /**
+   * Try to rewrite the key as LongType so we can use getLong(), if they key can fit with a long.
+   *
+   * If not, returns the original expressions.
+   */
+  private[joins] def rewriteKeyExpr(keys: Seq[Expression]): Seq[Expression] = {
```

Comment:
This will be a lot clear and less error prone if we can write the check first and go through the data twice, e.g.

```
if (keys.map(_.dataType.defaultSize).sum <= 8) {
  return keys
}

// do the rewrite here
```